pkg/build/paths.go

```diff
@@ -33,28 +30,68 @@ type Paths struct {
 	MetaDir string
 }
 
-func NewBuildPaths(root, executableName, zipName string) (Paths, error) {
+type pathsSettings struct {
+	// targetDir is used to calculate the default bin path.
+	targetDir string
+}
+
+// buildPathsSettings contains optional settings for build paths.
+type buildPathsSettings struct {
+	targetDir string
+}
+
+type BuildPathsOpt func(s *pathsSettings)
+
+func WithTargetDir(path string) BuildPathsOpt {
+	return func(s *pathsSettings) { s.targetDir = path }
+}
+
+func NewBuildPaths(root, executableName, zipName string, opts ...BuildPathsOpt) (Paths, error) {
 	var bp Paths
 	if !filepath.IsAbs(root) {
 		return bp, fmt.Errorf("root path %q is not absolute", root)
 	}
-	return bp.setDefaults(root, executableName, zipName), nil
+
+	settings := pathsSettings{}
+	// Apply options.
+	for _, o := range opts {
+		o(&settings)
+	}
+
+	return bp.setDefaults(root, executableName, zipName, settings), nil
 }
 
 func (bp Paths) ZipDir() string {
 	return filepath.Dir(bp.ZipPath)
 }
 
+// TargetDir is the absolute path to the dir where any other files
+// needed to be included in the zip file should be placed.
+func (bp Paths) TargetDir() string {
+	return filepath.Dir(bp.BinPath)
+}
+
 func (bp Paths) trimSpace() Paths {
 	// Placeholder func for consistency.
 	return bp
 }
 
-func (bp Paths) setDefaults(root, executableName, zipName string) Paths {
-	bp.WorkDir = root
-	bp.TargetDir = filepath.Join(root, Dirs.Target)
-	bp.BinPath = filepath.Join(root, Dirs.Target, executableName)
-	bp.ZipPath = filepath.Join(root, Dirs.Zip, zipName)
-	bp.MetaDir = filepath.Join(root, Dirs.Meta)
+func (bp Paths) setDefaults(root, executableName, zipName string, s pathsSettings) Paths {
+	if len(bp.WorkDir) == 0 {
+		bp.WorkDir = root
+	}
+	// Set the internal targetDir if it wasn't already set by an option.
+	if len(s.targetDir) == 0 {
+		s.targetDir = filepath.Join(bp.WorkDir, Dirs.Target)
+	}
+	if len(bp.BinPath) == 0 {
+		bp.BinPath = filepath.Join(s.targetDir, executableName)
+	}
+	if len(bp.ZipPath) == 0 {
+		bp.ZipPath = filepath.Join(bp.WorkDir, Dirs.Zip, zipName)
+	}
+	if len(bp.MetaDir) == 0 {
+		bp.MetaDir = filepath.Join(bp.WorkDir, Dirs.Meta)
+	}
```

Comment:
Currently only the `targetDir` and thus `BinPath` are externally settable, however I use a uniform pattern here so that the function will do what's expected (setting defaults where needed) if we allow other paths to be set externally in the future.